hoomd/operation.py

```diff
@@ -452,7 +452,13 @@ def _update_param_dict(self):
 
     @log(flag='state')
     def state(self):
-        self._update_param_dict()
+        try:
+            self._update_param_dict()
+        except AttributeError:
+            # attribute errors in _update_param_dict should not result in
+            # class has no attribute `state`
+            raise RuntimeError("Failed to update param dict")
```

Comment:
This change converts this error:
```
Traceback (most recent call last):
  File "test2.py", line 29, in <module>
    sim.write_debug_data('test.json')
  File "/scratch/joaander/build/hoomd/hoomd/simulation.py", line 248, in write_debug_data
    log = logger.log()
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in log
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 49, in dict_map
    new_dict[key] = func(value)
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in <lambda>
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 362, in __call__
    attr = getattr(self.obj, self.attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 94, in __getattr__
    raise AttributeError("Object {} has no attribute {}".format(
AttributeError: Object <class 'hoomd.md.nlist.Cell'> has no attribute state

```

Into this one that identifies the root cause:
```
Traceback (most recent call last):
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 456, in state
    self._update_param_dict()
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 451, in _update_param_dict
    self._param_dict[key] = getattr(self, key)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 90, in __getattr__
    return self._getattr_param(attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 377, in _getattr_param
    return getattr(self._cpp_obj, attr)
AttributeError: 'hoomd.md._md.NeighborListBinned' object has no attribute 'deterministic'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test2.py", line 29, in <module>
    sim.write_debug_data('test.json')
  File "/scratch/joaander/build/hoomd/hoomd/simulation.py", line 248, in write_debug_data
    log = logger.log()
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in log
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 49, in dict_map
    new_dict[key] = func(value)
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in <lambda>
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 362, in __call__
    attr = getattr(self.obj, self.attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 459, in state
    raise RuntimeError("Failed to update param dict")
RuntimeError: Failed to update param dict
```

Without this change, the `AttributeError` thrown by `_md.NeighborListBinned` causes Python to assume that `Cell` has no attribute `state`. I don't know if there is a more Pythonic way to do this.

I would guess that any `AttributeError` thrown in `_get_state` would also result in the same behavior so that call should possibly also be inside the try: block.

Reply:
This is fine. I have just noticed that properties can mask errors, and I hadn't found a good way to prevent that. I like this approach.